[AutoDiff] Fix `@differentiable` attribute SILGen and serialization.

dan-zheng opened the pull request:
- Propagate `@differentiable` attributes in `SILFunctionBuilder::addFunctionAttributes`.
  - Previously, this logic was missing, causing SIL `[differentiable]` attributes
    to be missing during SILGen.
  - This unblocks using VJP definitions in the stdlib (with some extra work).
- Serialize `AutoDiffParameterIndices` for `DifferentiableAttr` and use it as the
 primary source for differentiation parameter info (instead of
  `ParsedAutoDiffParameter`).
- Rework differentiation to never load SIL functions.
  - Previously, differentiation relied on explicit loading to get SIL
    `[differentiable]` attributes. That was a hack.
  - Now that `@differentiable` attribute is propagated and serialized,
    such loading is no longer necessary.
- Tighten differentiation infrastructure to uphold the following invariants:
  - Differentiable functions' primal/adjoint are visible only in defining module.
  - Differentiable functions' JVP/VJPs are always visible to other modules.
    The differentiation pass guarantees that JVP/VJPs exist.

There are two `TensorFlowRuntime` test regressions in GPE mode.
- Both are because used device set includes RUNTIME device in
  `GraphFunctionDeviceInfo::finalizeUsedDevices()`.

Todos:
- Replace adjoint definitions in stdlib with VJP definitions.
- Eliminate primal/adjoint from `@differentiable` attribute.
  Use JVP/VJPs everywhere.

dan-zheng committed:
[AutoDiff] Fix SILGen for `@differentiable` attribute.

Propagate `@differentiable` attributes in `SILFunctionBuilder::addFunctionAttributes`.
- Previously, this logic was missing, causing SIL `[differentiable]` attributes
  to be missing during SILGen.
- This unblocks using VJP definitions in the stdlib (with some extra work).

Serialize `AutoDiffParameterIndices` for `DifferentiableAttr` and use it as the
primary source for differentiation parameter info (instead of
`ParsedAutoDiffParameter`).

Fix `AutoDiff` tests.

There are two `TensorFlowRuntime` test regressions in GPE mode.
- Both are because used device set includes RUNTIME device in
  `GraphFunctionDeviceInfo::finalizeUsedDevices()`.

rxwei commented on include/swift/AST/AutoDiff.h:
```diff
@@ -109,12 +111,13 @@ class AutoDiffParameterIndices : public llvm::FoldingSetNode {
                                        ASTContext &C);
 
+private:
+  AutoDiffParameterIndices(llvm::SmallBitVector parameters)
```
```suggestion
  AutoDiffParameterIndices(const llvm::SmallBitVector &parameters)
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/21837/commits/6fe2c6458258a66d31c72730d89aed36f297996d.

rxwei commented on lib/SILOptimizer/Mandatory/TFDifferentiation.cpp:
```diff
@@ -990,6 +990,7 @@ class ADContext {
   lookUpOrRegisterDifferentiationTask(SILFunction *original,
                                       const SILAutoDiffIndices &indices,
                                       DifferentiationInvoker invoker) {
+    /*
```
You can just delete this hunk.

dan-zheng replied:
Woo! Done.

rxwei commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2460,11 +2459,11 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
-      switch (uncheckedWrtParams[i].getKind()) {
-      case AutoDiffParameter::Kind::Index: {
-        unsigned index = uncheckedWrtParams[i].getIndex();
+    for (size_t i = 0; i < parsedWrtParams.size(); i++) {
```
Some improvement:
```suggestion
    for (unsigned i : indices(parsedWrtParams)) {
```

dan-zheng replied:
Thanks! Done in https://github.com/apple/swift/pull/21837/commits/6fe2c6458258a66d31c72730d89aed36f297996d.

rxwei commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -2586,21 +2586,17 @@ ModuleFile::getDeclCheckedImpl(DeclID DID) {
-        }
-        // TODO: Deserialize CheckedParameterIndices.
+        llvm::SmallBitVector parametersBitVector(parameters.size());
+        for (unsigned i = 0; i < parameters.size(); i++)
```
```suggestion
        for (unsigned i : indices(parameters))
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/21837/commits/6fe2c6458258a66d31c72730d89aed36f297996d. Also done in `SerializeSIL.cpp` and `DeserializeSIL.cpp`.

rxwei commented on test/AutoDiff/derived_differentiable_properties.swift:
```diff
@@ -7,7 +7,7 @@ public struct Foo : Differentiable {
 
 // CHECK-AST-LABEL: public struct Foo : Differentiable {
-// CHECK-AST:   @sil_stored @differentiable()
+// CHECK-AST:   @sil_stored @differentiable(wrt: (self))
```
I love that things have a canonical printout now when you use checked parameter indices.

dan-zheng committed:
Address comments by @rxwei.

dan-zheng committed:
Merge branch 'tensorflow' into fix-differentiable-attr-silgen

dan-zheng committed:
Merge branch 'tensorflow' into fix-differentiable-attr-silgen

dan-zheng committed:
Minor fixes.

dan-zheng committed:
Merge branch 'tensorflow' into fix-differentiable-attr-silgen

dan-zheng commented:
@swift-ci Please clean test tensorflow macOS

dan-zheng commented:
@swift-ci Please clean test tensorflow macOS

dan-zheng commented:
@swift-ci Please clean test tensorflow macOS

dan-zheng commented:
@swift-ci Please test tensorflow Linux GPU

dan-zheng committed:
[AutoDiff] Fix `@differentiable` attribute SILGen and serialization.

- Propagate `@differentiable` attributes in `SILFunctionBuilder::addFunctionAttributes`.
  - Previously, this code was missing, causing `@differentiable` AST
    attributes to not propagate during SILGen.
- Rework differentiation to never load SIL functions.
  - Previously, differentiation relied on explicit loading to get SIL
    `[differentiable]` attributes. That was a hack.
  - Now that `@differentiable` attribute is propagated and serialized,
    such loading is no longer necessary.
- Tighten differentiation infrastructure to uphold the following invariants:
  - Differentiable functions' primal/adjoint are visible only in defining module.
  - Differentiable functions' JVP/VJP are always visible to other modules.
    The differentiation pass guarantees that JVP/VJPs exist.

Todos:
- Replace adjoint definitions in stdlib with VJP definitions.
- Eliminate primal/adjoint from `@differentiable` attribute.
  Use JVP/VJPs everywhere.

rxwei commented on lib/Serialization/Serialization.cpp:
```diff
@@ -2369,24 +2369,16 @@ void Serializer::writeDeclAttribute(const DeclAttribute *DA) {
+    auto paramIndices = attr->getParameterIndices();
+    assert(paramIndices && "Checked parameter indices must be resolved");
+    SmallVector<bool, 4> indices;
+    for (unsigned i = 0; i < paramIndices->parameters.size(); i++)
```
```suggestion
    for (unsigned i : indices(paramIndices->parameters.size()))
```

dan-zheng replied:
Thanks! Done in https://github.com/apple/swift/pull/21837/commits/5d8016b1850d755062fba2d9f19c12495569b5ac.

dan-zheng committed:
Add cross-module serialization test.

dan-zheng committed:
Address comments by @rxwei.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Fix `SILLoader` related issue.

Calling `module.getSILLoader()` here is necessary to prevent crashes related to
`lookUpFunctionInWitnessTable`. This might be a bug in the `SILLoader`
implementation.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Fix serialization tests.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Fix crasher during argument explosion optimization.

`test/AutoDiff/simple_model.swift` failed during argument epxlosion optimization.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow linux

dan-zheng commented:
@swift-ci Please test tensorflow linux

dan-zheng commented:
This was joint work with @rxwei. 🤝
